src/plots/cartesian/new_shape.js

```diff
@@ -0,0 +1,917 @@
+        displayOutlines(polygons, outlines, dragOptions, nCalls++);
+
+        dragOptions.isActiveShape = false; // i.e. to disable controllers
+        var shapes = addNewShapes(outlines, dragOptions);
+        if(shapes) {
+            Registry.call('_guiRelayout', gd, {
+                shapes: shapes // update active shape
+            });
+        }
+    }
+
+
+    // remove previous controllers - only if there is an active shape
+    if(gd._fullLayout._activeShapeIndex >= 0) clearOutlineControllers(gd);
+
+    var isActiveShape = dragOptions.isActiveShape;
+    var fullLayout = gd._fullLayout;
+    var zoomLayer = fullLayout._zoomlayer;
+
+    var dragmode = dragOptions.dragmode;
+    var isDrawMode = drawMode(dragmode);
+
+    if(isDrawMode) gd._fullLayout._drawing = true;
+
+    // make outline
+    outlines.attr('d', writePaths(polygons));
+
+    // add controllers
+    var rVertexController = MINSELECT * 1.5; // bigger vertex buttons
+    var vertexDragOptions;
+    var shapeDragOptions;
+    var indexI; // cell index
+    var indexJ; // vertex or cell-controller index
+    var copyPolygons;
+
+    copyPolygons = recordPositions([], polygons);
+
+    if(isActiveShape) {
+        var g = zoomLayer.append('g').attr('class', 'outline-controllers');
+        addVertexControllers(g);
+        addShapeControllers();
+    }
+
+    function startDragVertex(evt) {
+        indexI = +evt.srcElement.getAttribute('data-i');
+        indexJ = +evt.srcElement.getAttribute('data-j');
+
+        vertexDragOptions[indexI][indexJ].moveFn = moveVertexController;
+    }
+
+    function moveVertexController(dx, dy) {
+        if(!polygons.length) return;
+
+        var x0 = copyPolygons[indexI][indexJ][1];
+        var y0 = copyPolygons[indexI][indexJ][2];
+
+        var cell = polygons[indexI];
+        var len = cell.length;
+        if(pointsShapeRectangle(cell)) {
+            for(var q = 0; q < len; q++) {
+                if(q === indexJ) continue;
+
+                // move other corners of rectangle
+                var pos = cell[q];
+
+                if(pos[1] === cell[indexJ][1]) {
+                    pos[1] = x0 + dx;
+                }
+
+                if(pos[2] === cell[indexJ][2]) {
+                    pos[2] = y0 + dy;
+                }
+            }
+            // move the corner
+            cell[indexJ][1] = x0 + dx;
+            cell[indexJ][2] = y0 + dy;
+
+            if(!pointsShapeRectangle(cell)) {
+                // reject result to rectangles with ensure areas
+                for(var j = 0; j < len; j++) {
+                    for(var k = 0; k < cell[j].length; k++) {
+                        cell[j][k] = copyPolygons[indexI][j][k];
+                    }
+                }
+            }
+        } else { // other polylines
+            cell[indexJ][1] = x0 + dx;
+            cell[indexJ][2] = y0 + dy;
+        }
+
+        redraw();
+    }
+
+    function endDragVertexController(evt) {
+        Lib.noop(evt);
```

Comment:
?

Reply:
Good Call. Revised in https://github.com/plotly/plotly.js/pull/4775/commits/c5ff5dfeec859a1c46cfc1494155bab8a1270bd0.